drivers: ethernet: Fix STM32F7 Ethernet with DMA

alexanderwachter opened the pull request:
This PR fixes the Ethernet driver for STM32F7 MCUs with DMA.
Deactivating or invalidating the data cache didn't work. To fix the problem, the DMA buffers are moved to the DTCM (Data Tightly Coupled Memory). This new section is also introduced in this PR

Issue: #15910

Commit:
boards: arm: Activate DTCM for STM32F7 boards with Ethernet

Activate the DTCM for STM32F7 board that have Ethernet.
This is needed because the Ethernet driver puts the DMA buffer
to this section.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: ethernet: stm32: Put DMA buffer to DTCM section

For STM32F7 MCU the actual implementation doesn't work when the
DMA buffers are placed in the SRAM.
This might be a problem with caches.
To overcome this problem, the buffer is moved to the DTCM.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

erwango commented:
You need to rebase on https://github.com/zephyrproject-rtos/zephyr/pull/17120 to get CI pass

alexanderwachter commented:
Rebased.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -26,10 +26,22 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #include "eth_stm32_hal_priv.h"
 
+
+#if defined(DT_DTCM_BASE_ADDRESS)
```
Even if this is the only known way we currently have to get ethernet back on STM32F7, I'd prefer to see this as a Kconfig option. Eventually, we'll get the the __nocache option working as well and users will be able to choose the option they prefer.

alexanderwachter replied:
Done

erwango reviewed (changes requested):
Tested ok on nucleo_f746zg.
One request: move ethernet buffers under Kconfig option.

alexanderwachter commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -26,10 +26,27 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
+#if defined(CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER) && \
+    !defined(DT_DTCM_BASE_ADDRESS)
+#warning DTCM for DMA buffer is activated but DT_DTCM_BASE_ADDRESS is not present
```
Ok. Done

alexanderwachter commented:
@erwango I split the sections into data, bss and noinit to avoid unexpected behavior of uninitialized memory.

erwango commented on include/linker/sections.h:
```diff
@@ -56,6 +56,10 @@
 #define _CCM_BSS_SECTION_NAME		.ccm_bss
 #define _CCM_NOINIT_SECTION_NAME	.ccm_noinit
+
+#define _DTCM_DATA_SECTION_NAME	.dtcm_data
```
Wrong alignment here

alexanderwachter replied:
Done

erwango commented on soc/arm/st_stm32/stm32f7/dtcm.ld:
```diff
@@ -0,0 +1,40 @@
+/*
```
So far STM32F7 is indeed the only STM32 series, though for coherency (and "futureproofness"), please move this file under `st_stm32/common`.

alexanderwachter replied:
Done.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -26,10 +26,27 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#if defined(CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER) && \
+    defined(DT_DTCM_BASE_ADDRESS)
+static ETH_DMADescTypeDef
+	dma_rx_desc_tab[ETH_RXBUFNB] __aligned(4) __dtcm_noinit_section;
```
Is `__aligned(4)` required when you're defining section with `SUBALIGN(4)` in linker?

alexanderwachter replied:
Good point. I don't think so. DMA with nocache would need a 32 alignment but I couldn't find any restrictions for DMA and DTCM.

erwango reviewed (changes requested):
Last nitpicks, otherwise LGTM

aurel32 commented on dts/arm/st/f7/stm32f723.dtsi:
```diff
@@ -15,6 +15,11 @@
 	};
 
+	dtcm: memory@20000000 {
+		compatible = "st,stm32-dtcm";
```
The DTCM memory is not specific to the STM32F7 nor to ST, but is found in all Cortex-M7 SoCs. I therefore think we should probably call it something like 'arm,dtcm'.

alexanderwachter replied:
Where is this `Cortex-M7` exactly?

aurel32 reviewed (changes requested):
I think it's perfectly acceptable to use the DTCM memory for the Ethernet DMA buffers, however I am sure there is a way to get it working with normal memory by adding cache handling calls at the right places, and possibly putting the buffer descriptors in nocache memory. I don't have a STM32F7 board with an Ethernet device, so I can't really help there. However I really believe the commit message should be modified to not be so affirmative and clearly says it's a temporary solution that can be improved later.

About adding support for DTCM memory, the code looks all fine to me, but it should probably be moved at the Cortex-M7 level, it's nothing specific to STM32F7.

Commit:
linker: Add dtcm section for Cortex M7 MCUs

This commit adds a DTCM (Device Tightly Coupled Memory) section for
Cortex F7 MCUs. The Address and length is defined in the corresponding
device tree file.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
dts: stm32f7: Add Device Tightly Coupled Memory

This commit adds the DTCM to the STM32F7 MCUs

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
Updated according to comments from @aurel32.
DTCM is introduced for all Cortex M7 now. To activate it, the DTCM area has to be added to the MCUs device tree.

aurel32 reviewed (approved):
Thanks for the update, it looks fine to me. Just two minor things:
- The comment of the first commits says "Cortex F7" instead of "Cortex M7"
- I think that the DTCM memory could be enabled on all STM32F7 SoCs, not only the ones with Ethernet. DTCM can be useful for other things than Ethernet (e.g. data for a time critical algorithm). That can be done on a later commit/PR though.

alexanderwachter commented:
@aurel32 The DTCM is enabled on boards level. So far it's only activated on boards with Ethernet.
We could make another PR to enable all the boards. But then I would suggest to include the other vendors too.

sigvartmh commented:
Maybe try adding the linker stuff with some of these macros https://github.com/zephyrproject-rtos/zephyr/pull/15112

alexanderwachter commented:
@sigvartmh I tried that but there is no general CMake file for Cortex M7.

sigvartmh commented:
Maybe you could make one? Then it would be easier to maintain in the future if somethings changes or a change is required.

sigvartmh commented:
However it is not a blocker just a suggestion, feel free to ignore it.

pfalcon commented on dts/bindings/arm/arm,dtcm.yaml:
```diff
@@ -0,0 +1,17 @@
+
+# SPDX-License-Identifier: Apache-2.0
+title: DTCM
```
For reference, we have nxp,imx-dtcm.yaml/nxp,imx-itcm.yaml in the codebase. I understand it wouldn't be used here, so mostly a ping to @MaureenHelm to double-check if iMX's DTCM/ITCM are anyhow special, or it would make sense to use this newly added `arm,dtcm`.

pfalcon commented on dts/arm/st/f7/stm32f723.dtsi:
```diff
@@ -16,7 +16,7 @@
 
 	dtcm: memory@20000000 {
-		compatible = "st,stm32-dtcm";
+		compatible = "arm,dtcm";
```
Also cc: @MaureenHelm (and maybe @agansari) re: if it would make sense to switch iMX to generic arm,dtcm either.

agansari replied:
@pfalcon well, i.MX RT does not use 'standard' TCM memories, but dts should be the same. Let's see @MaureenHelm's feedback.

pfalcon reviewed (approved):
I don't see anything wrong from networking side, and overall looks pretty clear/clean, so give +1, following @erwango's review.

alexanderwachter commented on dts/arm/st/f7/stm32f769.dtsi:
```diff
@@ -16,7 +16,7 @@
 	};
 
 	dtcm: memory@20000000 {
-		compatible = "st,stm32-dtcm";
```
Clean now.
